src/engine/platform/opl4.cpp

```diff
@@ -0,0 +1,349 @@
+/**
```

Comment:
You based the OPL4 code off FM system? This might be a problem for macro handling...